src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Windows.cs

```diff
@@ -288,6 +289,30 @@ unsafe long GetFileLengthCore()
 
                 if (!Interop.Kernel32.GetFileInformationByHandleEx(this, Interop.Kernel32.FileStandardInfo, &info, (uint)sizeof(Interop.Kernel32.FILE_STANDARD_INFO)))
                 {
+                    if (this._path?.StartsWith(@"\\?\") ?? false)
+                    {
+                        byte[] buffer = ArrayPool<byte>.Shared.Rent(4096);
+                        bool success = Interop.Kernel32.DeviceIoControl(
+                            this,
+                            dwIoControlCode: Interop.Kernel32.IOCTL_STORAGE_READ_CAPACITY,
+                            lpInBuffer: IntPtr.Zero,
+                            nInBufferSize: 0,
+                            lpOutBuffer: buffer,
+                            nOutBufferSize: 4096,
+                            out _,
+                            IntPtr.Zero);
+
+                        if (!success)
+                        {
+                            throw Win32Marshal.GetExceptionForLastWin32Error(Path);
+                        }
+
+                        Span<byte> bufferSpan = new(buffer);
+                        success = MemoryMarshal.TryRead(bufferSpan, out Interop.Kernel32.StorageReadCapacity storageReadCapacity);
+                        Debug.Assert(success);
+                        return storageReadCapacity.DiskLength;
+                    }
+
```

Comment:
Suggest reversing the condition so that this throw can go directly after the pinvoke. That's because it's essential it rings directly after the pinvoke (as the error value can get overwritten) and that makes it clear to the reader that it does.

Reply:
Reversed